frontend/pages/api/auth/[...nextauth].ts

```diff
@@ -0,0 +1,369 @@
+const callbacks: CallbacksOptions = {};
+
+callbacks.signIn = async function signIn(user: User, provider, data) {
+  let accessToken: string;
+  let refreshToken: string;
+  let cookies: Array<string>;
+  if (provider.id === 'credentials') {
+    /*
+      Coming from the credentials provider, `data` is of this form:
+      {
+        csrfToken: 'example',
+        username: 'example',
+        password: 'example'
+      }
+    */
+    accessToken = user.accessToken;
+    refreshToken = user.refreshToken;
+    cookies = user.cookies;
+  } else {
+    let socialUser;
+    switch (provider.id) {
+      case 'discord': // https://next-auth.js.org/providers/discord
+        socialUser = {
+          id: data.id,
+          login: data.login,
+          name: data.name,
+          avatar: user.image
+        };
+        break;
+      case 'facebook': // https://next-auth.js.org/providers/facebook
+        socialUser = {
+          id: data.id,
+          login: data.login,
+          name: data.name,
+          avatar: user.image
+        };
+        break;
+      case 'github': // https://next-auth.js.org/providers/github
+        // TODO: https://modularhistory.atlassian.net/browse/MH-136
+        // https://getstarted.sh/bulletproof-next/add-social-authentication/5
+        // const emailRes = await fetch('https://api.github.com/user/emails', {
+        //   headers: {
+        //     'Authorization': `token ${account.accessToken}`
+        //   }
+        // })
+        // const emails = await emailRes.json()
+        // const primaryEmail = emails.find(e => e.primary).email;
+        // user.email = primaryEmail;
+        socialUser = {
+          id: data.id,
+          login: data.login,
+          name: data.name,
+          avatar: user.image
+        };
+        break;
+      case 'google': // https://next-auth.js.org/providers/google
+        socialUser = {
+          id: data.id,
+          login: data.login,
+          name: data.name,
+          avatar: user.image
+        };
+        break;
+      case 'twitter': // https://next-auth.js.org/providers/twitter
+        socialUser = {
+          id: data.id,
+          login: data.login,
+          name: data.name,
+          avatar: user.image
+        };
+        break;
+      default:
+        return false;
+    }
+    accessToken = accessToken || null; // TODO: https://modularhistory.atlassian.net/browse/MH-136: await getTokenFromYourAPIServer(account.provider, socialUser);
+    refreshToken = refreshToken || null; // TODO: https://modularhistory.atlassian.net/browse/MH-136: await getTokenFromYourAPIServer(account.provider, socialUser);
+    cookies = cookies || null;
+  }
+  user.accessToken = accessToken;
+  user.refreshToken = refreshToken;
+  user.cookies = cookies;
+  return true;
+};
+
+// https://next-auth.js.org/configuration/callbacks#jwt-callback
+callbacks.jwt = async function jwt(token, user?: User, account?, profile?, isNewUser?: boolean) {
```

Comment:
Do we not need types specified for account/profile? Would we want to add types for them eventually?

Reply:
Probably so. I'll take a look when I enable social auth in MH-136.